### PR TITLE
Test & fix for column uniqueness check

### PR DESCRIPTION
**Pull request by pavelgubin**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMetadataQuery.java [hunk 1]
```diff
@@ -537,6 +537,9 @@ public Boolean areColumnsUnique(RelNode rel, ImmutableBitSet columns) {
    */
   public Boolean areColumnsUnique(RelNode rel, ImmutableBitSet columns,
       boolean ignoreNulls) {
+    if (columns.length() > rel.getInputs().size()) {
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMetadataQuery.java]: Would you please add a comment describing an edge case?
From a first glance it is not clear why the condition is like that.

Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMetadataQuery.java]: Would you please add a comment describing an edge case?
From a first glance it is not clear why the condition is like that.

**Comment by pavelgubin** [comment]
Well, the validation I proposed first was overly optimistic. The idea was to check that _columns_ argument contains indices of columns that _rel_ can actually provide. But getInputs() is not good fit for this validation. So proposed different check close to the line where error actually occurs.

Error occurs at the following line inside _RelMdColumnUniqueness_:
`     
RexNode projExpr = projExprs.get(bit);
`

**Comment by vlsi** [comment]
>Error occurs at the following line inside RelMdColumnUniqueness:
>`RexNode projExpr = projExprs.get(bit);`

It is clear. It is not clear how execution gets there.
What causes `ImmutableBitSet columns` to have more bits than the projected columns?

**Comment by pavelgubin** [comment]
It happens in _RelBuilder.aggregate(GroupKey groupKey, Iterable<AggCall> aggCalls)_  call upper in stack:

This line:
`final ImmutableBitSet groupSet = ImmutableBitSet.of(registerExpressions(extraNodes, groupKey_.nodes));`
appends **new** nodes to _extraNodes_ collection if there are some expressions besides plain field references in groupKey_.nodes. Corresponding bits are set in _groupSet_.

Then at this line:
`final Boolean unique = mq.areColumnsUnique(peek(), groupSet);` we are trying to check uniqueness of these new columns on the underlying node given by peek() which was not changed.

**Comment by vlsi** [comment]
Then I disagree with the fix.

Note theres was `group sets contained expressions not in group key` error: https://github.com/vlsi/calcite/blob/1e7ae1c30a5e6f4335f2e4cc0059f2f2901e35fd/core/src/main/java/org/apache/calcite/tools/RelBuilder.java#L1095

I would rather add an assertion to `org.apache.calcite.rel.metadata.RelMdColumnUniqueness#areColumnsUnique` before `projExprs.get(bit)` to return meaningful error message in case of `ImmutableBitSet columns` is too large.

**Comment by pavelgubin** [comment]
I think assertion will not fix the problem and given exception already quite specific.

Mentioned error does not disallow expressions in group key, I'd say it assumes that expressions can be in group key but not in group sets. For example, it would be natural way of translate such SQL:
`select count(*) from emp group by substring(dept, 0, 5)`

Probably better to apply the fix to RelBuilder.aggregate() then to prevent it to check uniqueness if there are some additional expressions?

**Pull request closed without merging**